crates/sim2h/src/connection_mgr.rs

```diff
@@ -0,0 +1,319 @@
+        for _ in 0..100 {
+            if frame.is_none() {
+                frame = Some(WsFrame::default());
+            }
+            match wss.read(frame.as_mut().unwrap()) {
+                Ok(_) => {
+                    did_work = true;
+                    let data = frame.take().unwrap();
+                    debug!("socket {} read {} bytes", uri, data.as_bytes().len());
+                    if let Err(_) = evt_send.send(ConMgrEvent::ReceiveData(uri.clone(), data)) {
+                        debug!("socket evt channel closed {}", uri);
+                        // end task
+                        return;
+                    }
+                }
+                Err(e) if e.would_block() => break,
+                Err(e) => {
+                    error!("socket read error {} {:?}", uri, e);
+                    let _ = evt_send.send(ConMgrEvent::Disconnect(uri.clone(), Some(e.into())));
+                    // end task
+                    return;
+                }
+            }
+        }
+
+        // if we did work we might have more work to do,
+        // if not, let this task get parked for a time
+        if did_work {
+            tokio::task::yield_now().await;
+        } else {
+            tokio::time::delay_for(std::time::Duration::from_millis(5)).await;
+        }
+    }
+}
+
+/// internal actually spawn the above wss_task into the tokio runtime
+fn spawn_wss_task(uri: Lib3hUri, wss: TcpWss, evt_send: EvtSend) -> CmdSend {
+    let (cmd_send, cmd_recv) = tokio::sync::mpsc::unbounded_channel();
+    tokio::task::spawn(wss_task(uri, wss, evt_send, cmd_recv));
+    cmd_send
+}
+
+/// internal result enum for connection mgr processing loop
+enum ConMgrResult {
+    DidWork,
+    NoWork,
+    EndTask,
+}
+
+use ConMgrResult::*;
+
+/// internal loop for processing the connection mgr
+async fn con_mgr_task(mut con_mgr: ConnectionMgr, weak_ref_dummy: Weak<()>) {
+    loop {
+        if let None = weak_ref_dummy.upgrade() {
+            // no more references, let this task end
+            return;
+        }
+
+        match con_mgr.process() {
+            DidWork => tokio::task::yield_now().await,
+            NoWork => tokio::time::delay_for(std::time::Duration::from_millis(5)).await,
+            EndTask => return,
+        }
+    }
+}
+
+/// ConnectionMgr tracks a set of open websocket connections
+/// allowing you to send data to them and checking them for incoming data
+pub struct ConnectionMgr {
+    cmd_recv: CmdRecv,
+    evt_send_to_parent: EvtSend,
+    evt_send_from_children: EvtSend,
+    evt_recv_from_children: EvtRecv,
+    wss_map: std::collections::HashMap<Lib3hUri, CmdSend>,
+}
+
+impl ConnectionMgr {
+    /// spawn a new connection manager task, returning a handle for controlling it
+    /// and a receiving channel for any incoming data
+    pub fn new() -> (ConnectionMgrHandle, ConnectionMgrEventRecv) {
+        let (evt_p_send, evt_p_recv) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
One thing to keep in mind with unbounded channels is they could just keep growing. It doesn't allow any back pressure on the network and can be easy to miss.

Reply:
Totally agree, but bounded channels have their own problems until we've fully adopted futures because you have to await on the send side as well, lest you end up doing blocking inside your task and halting processing.